semantic-core/src/Analysis/ScopeGraph.hs

```diff
@@ -1,15 +1,130 @@
-( ScopeGraph
-, Entry(..)
+( ScopeGraph(..)
+, Decl(..)
+, scopeGraph
+, scopeGraphAnalysis
 ) where
 
+import           Analysis.Eval
+import           Analysis.FlowInsensitive
+import           Control.Applicative (Alternative (..))
+import           Control.Effect.Carrier
+import           Control.Effect.Fail
+import           Control.Effect.Fresh
+import           Control.Effect.Reader
+import           Control.Effect.State
+import           Control.Monad ((>=>))
+import           Data.File
+import           Data.Foldable (fold)
+import           Data.Function (fix)
+import           Data.List.NonEmpty
 import           Data.Loc
 import qualified Data.Map as Map
+import           Data.Name
+import           Data.Proxy
 import qualified Data.Set as Set
+import           Data.Text (Text)
+import           Data.Traversable (for)
+import           Prelude hiding (fail)
 
-data Entry = Entry
-  { entrySymbol :: String -- FIXME: Text
-  , entryLoc    :: Loc
+data Decl = Decl
+  { declSymbol :: Text
+  , declLoc    :: Loc
   }
+  deriving (Eq, Ord, Show)
 
-type ScopeGraph = Map.Map Entry (Set.Set Entry)
+newtype Ref = Ref Loc
+  deriving (Eq, Ord, Show)
+
+newtype ScopeGraph = ScopeGraph { unScopeGraph :: Map.Map Decl (Set.Set Ref) }
+  deriving (Eq, Ord, Show)
+
+instance Semigroup ScopeGraph where
+  ScopeGraph a <> ScopeGraph b = ScopeGraph (Map.unionWith (<>) a b)
+
+instance Monoid ScopeGraph where
+  mempty = ScopeGraph Map.empty
+
+scopeGraph
+  :: Ord term
+  => (forall sig m
+     .  (Carrier sig m, Member (Reader Loc) sig, MonadFail m)
+     => Analysis term User ScopeGraph m
+     -> (term -> m ScopeGraph)
+     -> (term -> m ScopeGraph)
+     )
+  -> [File term]
+  -> (Heap User ScopeGraph, [File (Either (Loc, String) ScopeGraph)])
+scopeGraph eval
+  = run
+  . runFresh
+  . runHeap
+  . traverse (runFile eval)
+
+runFile
+  :: ( Carrier sig m
+     , Effect sig
+     , Member Fresh sig
+     , Member (State (Heap User ScopeGraph)) sig
+     , Ord term
+     )
+  => (forall sig m
+     .  (Carrier sig m, Member (Reader Loc) sig, MonadFail m)
+     => Analysis term User ScopeGraph m
+     -> (term -> m ScopeGraph)
+     -> (term -> m ScopeGraph)
+     )
+  -> File term
+  -> m (File (Either (Loc, String) ScopeGraph))
+runFile eval file = traverse run file
+  where run = runReader (fileLoc file)
+            . runReader (Map.empty @User @Loc)
+            . runFailWithLoc
+            . fmap fold
+            . convergeTerm (Proxy @User) (fix (cacheTerm . eval scopeGraphAnalysis))
+
+-- FIXME: decompose into a product domain and two atomic domains
+scopeGraphAnalysis
+  :: ( Alternative m
+     , Carrier sig m
+     , Member (Reader Loc) sig
+     , Member (Reader (Map.Map User Loc)) sig
+     , Member (State (Heap User ScopeGraph)) sig
+     )
+  => Analysis term User ScopeGraph m
+scopeGraphAnalysis = Analysis{..}
+  where alloc = pure
+        bind name _ m = do
+          loc <- ask @Loc
+          local (Map.insert name loc) m
```

Comment:
Binding remembers (locally) the location for the binding site. Unfortunately, this will generally be the location (and especially the `Span`) of the `Ann` node surrounding the `Lam` or `:>>=`, making it a little imprecise. This could be improved by adding (optional) locations to these nodes for the bound variables, and since Core is curried, we’d do the right thing for multiple parameters automatically.

Reply:
👍 File an issue about this?

Reply:
I’m actually just tackling it in a follow-up PR.